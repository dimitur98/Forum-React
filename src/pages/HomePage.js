import React, { Component } from 'react';
import '../App.css';
import Header from '../components/header'
import Footer from '../components/footer'
import Category from '../components/category'

class HomePage extends Component{
  constructor(props){
    super(props)

    this.state = {
      categories: []
    }
  }

  getAllNotDeletedCategories = async() => {
    const promise = await fetch('http://localhost:9999/api/category/allCategories')
    const categories = await promise.json()
    console.log(categories)
    this.setState({
      categories
    })
  }

  renderCategories(){
    const { categories } = this.state
    
    return categories.map((category) => {
      return(
        <Category key = {category._id } name = {category.name} imageUrl = {category.imageUrl} id = {category._id}/>
      )
    })
  }

  componentDidMount(){
    this.getAllNotDeletedCategories()
  }

  render(){
    return (
      <div>     
          <Header />
          <div class="container">
              <main role="main" class="pb-3">
                  <div class="text-center">
                      <h1 class="display-3">ForumSystem</h1>

                      <a asp-controller="Posts" asp-action="Create" class="btn btn-primary btn-lg">Create new forum post</a>
                  </div>
                  <hr />
                  <div class="row">
                      {this.renderCategories()}                             
                  </div>
              </main>
          </div>
          <Footer />     
      </div>
    )
  }
}

export default HomePage;
