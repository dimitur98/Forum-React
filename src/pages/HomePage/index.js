import React, { Component } from 'react';
import '../../App.css';
import { Link } from 'react-router-dom'
import Category from '../../components/category'
import UserContext from '../../Context'
import PageWrapper from '../../components/pageWrapper'

class HomePage extends Component{
  constructor(props){
    super(props)

    this.state = {
      categories: [],
      admin: true,
      counter:0
    }
  }
  static contextType = UserContext
  getAllNotDeletedCategories = async() => {
    const promise = await fetch('http://localhost:9999/api/category/allCategories')
    const categories = await promise.json()
    this.setState({
      categories
    })
  }

  renderCategories(){
    const { categories } = this.state
    var counter = 0
    return categories.map((category) => {
      counter += 1
      return(
        <Category 
          refresh = {this.refresh} 
          key = {category._id } 
          name = {category.name} 
          imageUrl = {category.imageUrl} 
          id = {category._id}
          testId = {counter}
          />
      )
    })
  }

  isAdmin=()=>{
    const { user } = this.context
    if(user){
      if(user.role === 'admin'){
        this.setState({admin: true})
        return
      }
    }
    this.setState({admin:false})
  }

  refresh = () => {
    this.getAllNotDeletedCategories()
  }

  componentDidMount(){
    this.getAllNotDeletedCategories()
    this.isAdmin()
  }

  render(){
    const {admin} = this.state
    const {loggedIn} = this.context
    return (
      <PageWrapper title='Home - DForum'>
          <div className="container">
              <main role="main" className="pb-3">
                  <div className="text-center">
                      <h1 className="display-3">DForum</h1>

                      {(loggedIn && admin) && <Link to='/CreateCategory' className="btn btn-primary btn-lg">Add new category</Link>}
                  </div>
                  <hr />
                  <div className="row">
                      {this.renderCategories()}                             
                  </div>
              </main>
          </div>
      </PageWrapper>
    )
  }
}

export default HomePage;
