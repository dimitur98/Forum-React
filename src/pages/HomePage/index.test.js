import React from 'react'
import Enzyme from 'enzyme';
import renderer from 'react-test-renderer'
import HomePage from '.'
import TestingEnvironment from '../../test-utils/router'
import {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
 
Enzyme.configure({ adapter: new Adapter() });
jest.mock('../../components/pageWrapper', ()=>'pageWrapper')

describe('Home page', ()=>{
    it("should show button 'Add category' only for admins", () =>{
        const tree = renderer.create(
            <TestingEnvironment value={{
                user:{
                    email: "test",
                    role:'admin'
                },
                loggedIn: true
            }}>
                <HomePage/>
            </TestingEnvironment>
        ).toJSON()
        expect(tree).toMatchSnapshot()
    })
    it("should not show button 'Add category'", () =>{
        const tree = renderer.create(
            <TestingEnvironment value={{
                user:null,
                loggedIn: false
            }}>
                <HomePage/>
            </TestingEnvironment>
        ).toJSON()
        expect(tree).toMatchSnapshot()
    })
    it("should get categories",async()=>{
        const wrapper = shallow(<HomePage />);
        const instance = wrapper.instance();
        await instance.getAllNotDeletedCategories()
        expect(wrapper.state('categories').length).toBeGreaterThan(0);
    })
})