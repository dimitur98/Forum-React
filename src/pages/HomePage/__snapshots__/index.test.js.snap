// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page should not show button 'Add category' 1`] = `
<pageWrapper
  isAdmin={[Function]}
>
  <div
    className="container"
  >
    <main
      className="pb-3"
      role="main"
    >
      <div
        className="text-center"
      >
        <h1
          className="display-3"
        >
          DForum
        </h1>
      </div>
      <hr />
      <div
        className="row"
      />
    </main>
  </div>
</pageWrapper>
`;

exports[`Home page should show button 'Add category' only for admins 1`] = `
<pageWrapper
  isAdmin={[Function]}
>
  <div
    className="container"
  >
    <main
      className="pb-3"
      role="main"
    >
      <div
        className="text-center"
      >
        <h1
          className="display-3"
        >
          DForum
        </h1>
        <a
          class="btn btn-primary btn-lg"
          href="/CreateCategory"
          onClick={[Function]}
        >
          Add new category
        </a>
      </div>
      <hr />
      <div
        className="row"
      />
    </main>
  </div>
</pageWrapper>
`;
